### DISCUSSION TITLE
Creating signed tags in a Github action.

### DISCUSSION THREAD

**Discussion by jmonguilo** [community/community#27016] [Actions]
<p>Hello,</p>
<p>I have a question of how to handle the Github Actions environment.</p>
<p>In one of my repos, I am trying to automate the creation of release tags when merging to master.</p>
<p>My current approach consists of using octokit/rest.js to access Github when an action is triggered: When a pull request to master has been closed and merged, the action executes a Node file which, using a personal Github token, calls git.createTag and git.createRef, which creates the Tags. In a nutshell:</p>
<p>My Github action:</p>
<pre><code>on:
   pull_request:
       ...

jobs:
...
uses: my_actions/generate-tags@v1
with:
token: ${{ secrets.MY_TOKEN }}
otherParams: otherParams
</code></pre>
<p>My js file:</p>
<pre><code>...  
const github = require('@actions/github');
...  
const oct =new github.GitHub(MY_TOKEN);
octokit.git.createTag(params).then(// SAVE tag_sha);
octokit.git.createRef(tag_sha);
</code></pre>
<p>This works. However, I would like the Tags to be signed (verified) as well. Is it there any  approach where tags (or any action in general) clould be signed within a Github action environment? How can/should I provide a PGP key to octokit/rest.js to sign? Should I use some bash commands instead/before?</p>
<p>Thank you!</p>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/jmonguilo">@jmonguilo</a> , </p>
<p>I tested this action : mathieudutour/github-tag-action . I can automatically bump and tag master, on merge, with the latest SemVer formatted version.  And the tag created by this action has “Verified” tag.</p>
<p>The workflow is triggered by pull_request not push event. </p>
<p></p><div class="lightbox-wrapper"><a class="lightbox" href="https://aws1.discourse-cdn.com/github/original/2X/d/dfe862d17faee37817f0c90c9f1cfe9debfabb5f.png" data-download-href="/uploads/short-url/vWMmmihuqGb4zR5euRw9jeflWzR.png?dl=1" title="verified.png" rel="nofollow noopener"><img src="https://user-images.githubusercontent.com/24806478/181101162-6c253a87-7492-4273-ba41-7366bb819eed.png" alt="verified.png" data-base62-sha1="vWMmmihuqGb4zR5euRw9jeflWzR" width="690" height="201" srcset="https://aws1.discourse-cdn.com/github/optimized/2X/d/dfe862d17faee37817f0c90c9f1cfe9debfabb5f_2_690x201.png, https://aws1.discourse-cdn.com/github/original/2X/d/dfe862d17faee37817f0c90c9f1cfe9debfabb5f.png 1.5x, https://aws1.discourse-cdn.com/github/original/2X/d/dfe862d17faee37817f0c90c9f1cfe9debfabb5f.png 2x" data-small-upload="https://aws1.discourse-cdn.com/github/optimized/2X/d/dfe862d17faee37817f0c90c9f1cfe9debfabb5f_2_10x10.png"><div class="meta">
<svg class="fa d-icon d-icon-far-image svg-icon" aria-hidden="true"><use xlink:href="#far-image"></use></svg><span class="filename">verified.png</span><span class="informations">1009×294 9.12 KB</span><svg class="fa d-icon d-icon-discourse-expand svg-icon" aria-hidden="true"><use xlink:href="#discourse-expand"></use></svg>
</div></a></div>
<p>There is the ts file of the action: <a href="https://github.com/mathieudutour/github-tag-action/blob/master/src/main.ts" rel="nofollow noopener">https://github.com/mathieudutour/github-tag-action/blob/master/src/main.ts</a></p>
<p>If this could not help, could  you share your action here? </p>

**Comment by jmonguilo** [comment]
<p>Hi <a class="mention" href="/yanjingzhu">@yanjingzhu</a>, thank you for your reply!.</p>
<p>I am trying to use a personal access token instead of Github’s token. I still need to figure out how to pass signature information as well.</p>
<p>Nevertheless, I have also tried to use GITHUB_TOKEN instead of MY_TOKEN, and could not get the verified tag. Perhaps something different in my approach with respect?</p>
<p>Here is the complete code I am using to test the action (putting manually tag names fot the moment, but the interaction with octokit at the end is the same as mathieudutour/github-tag-action):</p>
<p>./.github/workflows/release.yml</p>
<pre><code>name: Release Pull Request

on:
  pull_request:

jobs:
  signed-commit:
    runs-on: ubuntu-latest
    steps:
      - uses: actions/checkout@v2
      - uses: actions/setup-node@v1
      - name: Tags
        uses: ./tag
        with:
          token: ${{ secrets.MY_TOKEN }}
          tag: '2020/04/28'
          message: 'Release test'
          commit: ${{ github.event.pull_request.head.sha }}
</code></pre>
<p>./tag/index.js action:</p>
<pre><code>const cp = require('child_process');

cp.execSync(`cd ${__dirname}; yarn install`);

const core = require('@actions/core');
const github = require('@actions/github');

async function tag() {

    const token = core.getInput("token", { required: true });
    const tagName = core.getInput("tag", { required: true });
    const tagMessage = core.getInput("message", { required: true });
    const [repoOwner, repoName] = process.env.GITHUB_REPOSITORY.split("/");

    const oct = new github.GitHub(token);

    const createdTag = await oct.git.createTag({
        owner: repoOwner,
        repo: repoName,
        tag: tagName,
        message: tagMessage,
        object: process.env.GITHUB_SHA,
        type: "commit"
    });

    if (createdTag.status !== 201) {
        core.setFailed(`Could not create tag. Received ${createdTag.status} from API`)
    }

    await oct.git.createRef({
        owner: repoOwner,
        repo: repoName,
        ref: 'refs/tags/' + tagName,
        sha: createdTag.data.sha
    }).then(
        ({ status }) =&gt; {
            if (status !== 201) {
                core.setFailed(`Could not create reference. Received ${status} from API`)
            }
        }
    )
}

tag().catch(
    (err) =&gt; {
        core.setFailed(err.message)
    }
);
</code></pre>

**Comment by Yanjingzhu** [comment]
<p>Hi <a class="mention" href="/jmonguilo">@jmonguilo</a> , </p>
<p>Thank you for sharing your index.js file . I have tested it in my action. The tag created by this action doesn’t have a verified symbol. I am trying to ask for help from github engineering team. This may take sometime. Appreciate your patience.</p>

**Comment by Yanjingzhu** [comment, marked as answer]
<p><a class="mention" href="/jmonguilo">@jmonguilo</a>  I got response from engineering team. </p>
<p>In your action, new tag is created using REST API, there is no method to create a signed tag using REST API.</p>
<p>The action <a href="https://github.com/mathieudutour/github-tag-action" rel="nofollow noopener">mathieudutour/github-tag-action</a> I provided before , it does the signing of the commit on the command line and then pushes it up. </p>
<p>We would encourage you to customize github-tag-action instead of using octokit directly.  </p>

**Comment by AraHaan** [comment]
<p>nice, now actions can do that now. Sweet.</p>

**Comment by jmonguilo** [comment]
<p>Hi <a class="mention" href="/yanjingzhu">@yanjingzhu</a> </p>
<p>Thank you very much, indeed I think I will use this approach in the future.</p>
<p>On the mean time, I have used Octokit’s create release</p>
<pre><code>oct.repos.createRelease
</code></pre>
<p>which for this particular case its acceptable and creates a signed tag with the committer signature.</p>

**Comment by seunggabi** [comment]
I recommend this.

https://github.com/seunggabi/auto-close-milestone/blob/main/.github/workflows/auto-close-milestone.yaml
